File: DesignPatterns/src/C++/002-GOFPatterns/002-StructuralPatterns/002-Flyweight/IntValue/IntValue.cpp
#include "IntValue.h"

std::map<int, std::shared_ptr<IntValue>> IntValue::ms_cache;

IntValue::IntValue(int val) : m_val{} {}

std::shared_ptr<IntValue> IntValue::value_of(int val)
{
    if (val < -128 || val > 127)
        return std::shared_ptr<IntValue>{new IntValue{val}};

    if (ms_cache.find(val) == ms_cache.end())
        ms_cache[val] = std::shared_ptr<IntValue>{new IntValue{val}};

    return ms_cache[val];
}

File: src/DesignPatterns/src/Java/002-GOF-Patterns/002-StructuralPatterns/002-Flywieight/IntValue/src/org/csystem/app/IntValue.java
package org.csystem.app;

public class IntValue {
    private final int m_val;
    private static IntValue [] ms_cache = new IntValue[256];

    private IntValue(int val)
    {
        m_val = val;
    }

    public int getVal()
    {
        return m_val;
    }

    public static IntValue valueOf(int val) //Factory method
    {
        if (val < -128 || val > 127)
            return new IntValue(val);

        if (ms_cache[val + 128] == null)
            ms_cache[val + 128] = new IntValue(val);

        return ms_cache[val + 128];
    }
}
